### ISSUE TITLE
Bug: Issue with `noUncheckedIndexedAccess` and `no-unnecessary-type-assertion`

### ISSUE THREAD

**Issue by Stadly** [typescript-eslint/typescript-eslint#11343]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

```ts
const users: Record<string, string> = {};
export const getUser = (name: string) => {
  users[name] ??= 'hei';
  return users[name]!; // Error: @typescript-eslint/no-unnecessary-type-assertion
};

const groups: Record<string, string> = {};
export const getGroup = (name: string) => {
  name = `${name}-grp`;
  groups[name] ??= 'hei';
  return groups[name]!; // No error!
};
```

I expected that the `return` from both functions should produce the `@typescript-eslint/no-unnecessary-type-assertion` error. Only the first does. If I remove `noUncheckedIndexedAccess`, I get the expected behavior.

### Reproduction Repository Link

https://stackblitz.com/edit/vitejs-vite-imvtqxsz?file=src%2Ffoo.ts

### Repro Steps

1) Open the StackBlitz
2) `npx eslint .`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.35.0` |
| `@typescript-eslint/parser`            | `8.35.0` |
| `@typescript-eslint/scope-manager`     | `8.35.0` |
| `@typescript-eslint/typescript-estree` | `8.35.0` |
| `@typescript-eslint/type-utils`        | `8.35.0` |
| `@typescript-eslint/utils`             | `8.35.0` |
| `TypeScript`                           | `5.8.3` |
| `ESLint`                               | `9.29.0` |
| `node`                                 | `20.19.1` |

**Comment by kirkwaiblinger** [comment]
Hi @Stadly!

Here's a simplified reproduction that demonstrates why this is working correctly. I'm not sure why TS narrows the type differently, but, according to how TS narrows the types, the rule is accurately reporting whether a type assertion is redundant or not.

[Playground](https://typescript-eslint.io/play/#ts=5.8.3&showAST=scope&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3jAXPASuEjMADyYwCWKA5gDTwXU0B8A3AFAcBmArimAyVU8bkiQAKNHia0AlDIxVa8AN4d42ANpoAuvAD8BgLzwA5GKRnOmyt3gSAslAwALAHQwoKYEgC2EnLwLPAADO4ArEHqmpoA9HHwAO7EANZoGrFwGLwwKNp6NvAAvvAgEGgIMbEJjEjwbpRo8FAQcFDAAJ7wrlDNGJ0ADgiy9C0%2BDa4gcGJwkwh9lTBCIk3w-CjgIGhosJ3umZrZufkwOroAhEXFHDc8-ILC%2BQBGsFKKyjQKjErMapnNUxmND%2BECNWjWTKZM56QwmcyWSG2exOFweLw%2BfyBYJhSLRQ7wWq%2BbYoMxYFIwVJsHp9BpDEa-FQAH3WPhA3GoIGABOOeQKumuZQqVQJtRBkzWi2mK3ya02YG2uxg%2Bx5YJO-KumRuxSAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UAO4jl4MAF8Q2oA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false)

### Code

```ts
declare const r: Record<string, string>;

function foo(s: string): string {
  r[s] ??= 'foo';
  if (Math.random() > 0.5) {
    // works
    return r[s];
  } else {
    // so this already has type string, and therefore the assertion is unnecessary.
    return r[s]!;
  }
}

function bar(s: string): string {
  s = 'something';
  
  r[s] ??= 'foo';
  if (Math.random() > 0.5) {
    // doesn't work; has type string | undefined
    return r[s];
  } else {
    // so this assertion is necessary.
    return r[s]!;
  }
}
```

### Eslint config

```json
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "warn"
  }
}
```

### TypeScript config

```json
{
  "compilerOptions": {
    "strictNullChecks": true,
    "noUncheckedIndexedAccess": true
  }
}
```

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `8.35.0` |
| `@typescript-eslint/parser` | `8.35.0` |
| `TypeScript` | `5.8.3` |
| `ESLint` | `9.26.0` |
| `node` | `web` |

**Comment by Stadly** [comment]
Thanks @kirkwaiblinger! For future reference, here is a reproduction in the Typescript playground: https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true#code/CYUwxgNghgTiAEYD2A7AzgF3jAXPASuEjMADyYwCWKA5gDTwXU0B8A3AFAcBmArimAyVU8bkiQAKANYgAnnia0AlAoxVa8AN4d48GbPgBeeAHI0SALYgMAC2Ym28APRP4AYUtWUWJLyy3KNAYoFGB4WwQQGBhieGBAqAAHRJBYNABCLl0YAG19AF14AH4i4xMAI1gHHWxrXhgUbDy5fMcXeABRaOI8ABVZFNNFGngAH3h+UG5qEGATeED4FCQsKDQ0ShoUKHKIBAwkcIGEMzV7ADoOAF8gA

**Comment by kirkwaiblinger** [comment]
@Stadly Probably worth filing an issue with TS or asking on their discord why this case behaves this way if you want to pursue this further 👍
